File: test/php/widget/PaymentTest.php
<?php

/**
 * @fileoverview Dana Widget Payment Test Suite for DANA Widget Integration
 * 
 * This suite validates the DANA widget payment functionality of the DANA Widget API.
 * Each test case includes comprehensive documentation following PHPDoc standards.
 * Tests cover success scenarios, error handling, validation, and idempotency behavior.
 * 
 * @author Integration Test Team
 * @version 1.0.0
 * @since 2024
 * @package DanaUat\Widget
 * @requires dana-php-sdk
 * @requires phpunit
 * @requires dotenv
 * 
 * Test Categories:
 * - Payment Success: Validates successful payment processing
 * - Error Handling: Tests various error scenarios and validation
 * - Idempotency: Verifies duplicate request handling behavior
 * - Security: Tests signature validation and authentication
 * - Timeout: Validates timeout handling
 */

namespace DanaUat\Widget;

use PHPUnit\Framework\TestCase;
use Dana\Widget\v1\Api\WidgetApi;
use Dana\Configuration;
use Dana\ObjectSerializer;
use Dana\Env;
use Dana\ApiException;
use DanaUat\Helper\Assertion;
use DanaUat\Helper\Util;
use DanaUat\Widget\PaymentUtil;
use Exception;

class PaymentTest extends TestCase
{
    private static $titleCase = 'Payment';
    private static $jsonPathFileWidget = 'resource/request/components/Widget.json';
    private static $apiInstanceWidget;
    private static $configuration;
    private static $merchantId;
    private static $sharedOriginalPartnerReference;

    public static function setUpBeforeClass(): void
    {
        // Set up configuration with authentication settings
        self::$configuration = new Configuration();

        // The Configuration constructor automatically loads values from environment variables
        // But we can manually set them if needed
        self::$configuration->setApiKey('PRIVATE_KEY', getenv('PRIVATE_KEY'));
        self::$configuration->setApiKey('ORIGIN', getenv('ORIGIN'));
        self::$configuration->setApiKey('X_PARTNER_ID', getenv('X_PARTNER_ID'));
        self::$configuration->setApiKey('ENV', Env::SANDBOX);

        self::$merchantId = getenv('MERCHANT_ID');
        self::$apiInstanceWidget = new WidgetApi(null, self::$configuration);

        // Create one paid order (runs automation once); cache is reused by QueryOrderTest etc.
        self::$sharedOriginalPartnerReference = PaymentUtil::createPaymentWidgetPaid("PaymentSuccess");
    }

    /**
     * @testdox Should give success response for payment scenario
     * @description Validates successful payment processing through DANA Widget API
     * @test Verifies that a valid payment request returns success response with proper reference number
     * @expectedResult Payment request succeeds and returns valid response object
     * @category PaymentSuccess
     * @priority High
     * @author Integration Test Team
     * @since 1.0.0
     */
    public function testPaymentSuccess(): void
    {
        Util::withDelay(function () {
            $caseName = 'PaymentSuccess';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );
            $jsonDict['partnerReferenceNo'] = PaymentUtil::generatePartnerReferenceNo();

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            $apiResponse = self::$apiInstanceWidget->widgetPayment($requestObj);
            // Assert the response matches the expected data
            Assertion::assertResponse(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName,
                $apiResponse->__toString(),
                []
            );
        });
    }

    /**
     * @testdox Should give fail response for merchant not exist or status abnormal scenario
     * @description Validates error handling when merchant ID does not exist or has abnormal status
     * @test Verifies that invalid merchant scenarios return appropriate error responses
     * @expectedResult API returns merchant-related error with proper error code and message
     * @category ErrorHandling
     * @priority High
     * @author Integration Test Team
     * @since 1.0.0
     */
    public function testPaymentFailMerchantNotExistOrStatusAbnormal(): void
    {
        Util::withDelay(function () {
            $caseName = 'PaymentFailMerchantNotExistOrStatusAbnormal';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );

            $jsonDict['merchantId'] = self::$merchantId;
            $jsonDict['partnerReferenceNo'] = PaymentUtil::generatePartnerReferenceNo();
            $jsonDict['validUpTo'] = Util::generateFormattedDate(600, 7);

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            // Assert the response matches the expected data
            try {
                self::$apiInstanceWidget->widgetPayment($requestObj);
            } catch (ApiException $e) {
                Assertion::assertApiException(
                    self::$jsonPathFileWidget,
                    self::$titleCase,
                    $caseName,
                    $e,
                    [],
                    $jsonDict
                );
            }
        });
    }

    /**
     * @testdox Should give fail response for inconsistent payment scenario
     * @description Validates detection of inconsistent payment data with same partner reference
     * @test Verifies that modified payment data with same reference ID is properly rejected
     * @expectedResult API returns inconsistent request error when data differs for same reference
     * @category DataValidation
     * @priority High
     * @author Integration Test Team
     * @since 1.0.0
     */
    public function testPaymentFailInconsistentRequest(): void
    {
        $this->markTestSkipped('Inconsistent request scenario skipped');
        Util::withDelay(function () {
            $caseName = 'PaymentFailInconsistentRequest';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );

            $jsonDict['merchantId'] = self::$merchantId;
            $fixedPartnerRef = PaymentUtil::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $fixedPartnerRef;
            $jsonDict['validUpTo'] = Util::generateFormattedDate(600, 7);

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            try {
                // First request with original data
                self::$apiInstanceWidget->widgetPayment($requestObj);

                // Second request with modified amount but same partnerReferenceNo
                $jsonDict['amount']['currency'] = 'IDR';
                $jsonDict['amount']['value'] = '2000.00';
                $modifiedRequestObj = ObjectSerializer::deserialize(
                    $jsonDict,
                    'Dana\Widget\v1\Model\WidgetPaymentRequest'
                );

                self::$apiInstanceWidget->widgetPayment($modifiedRequestObj);
            } catch (ApiException $e) {
                Assertion::assertApiException(
                    self::$jsonPathFileWidget,
                    self::$titleCase,
                    $caseName,
                    $e,
                    [],
                    $jsonDict
                );
            }
        });
    }

    /**
     * @testdox Should give fail response for invalid field format scenario
     * @description Validates error handling for invalid field formats in payment requests
     * @test Verifies that malformed or invalid field values return appropriate validation errors
     * @expectedResult API returns field format validation error with descriptive message
     * @category FieldValidation
     * @priority Medium
     * @author Integration Test Team
     * @since 1.0.0
     */
    public function testPaymentInvalidFieldFormat(): void
    {
        Util::withDelay(function () {
            $caseName = 'PaymentFailInvalidFormat';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );

            $jsonDict['merchantId'] = self::$merchantId;
            $jsonDict['partnerReferenceNo'] = PaymentUtil::generatePartnerReferenceNo();
            $jsonDict['validUpTo'] = Util::generateFormattedDate(600, 7);

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            // Assert the response matches the expected data
            try {
                self::$apiInstanceWidget->widgetPayment($requestObj);
            } catch (ApiException $e) {
                Assertion::assertApiException(
                    self::$jsonPathFileWidget,
                    self::$titleCase,
                    $caseName,
                    $e,
                    [],
                    $jsonDict
                );
            }
        });
    }

    /**
     * @testdox Should give fail response for missing or invalid mandatory field scenario
     * @description Validates error handling when required fields are missing or invalid
     * @test Verifies that requests without mandatory headers (X-TIMESTAMP) are rejected
     * @expectedResult API returns mandatory field validation error with clear indication
     * @category MandatoryFieldValidation
     * @priority High
     * @author Integration Test Team
     * @since 1.0.0
     */
    public function testPaymentFailMissingOrInvalidMandatoryField(): void
    {
        Util::withDelay(function () {
            $caseName = 'PaymentFailMissingOrInvalidMandatoryField';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );

            $jsonDict['merchantId'] = self::$merchantId;
            $jsonDict['partnerReferenceNo'] = PaymentUtil::generatePartnerReferenceNo();
            $jsonDict['validUpTo'] = Util::generateFormattedDate(600, 7);

            // Create headers without timestamp to test validation
            $headers = Util::getHeadersWithSignature(
                'POST',
                '/rest/redirection/v1.0/debit/payment-host-to-host.htm',
                $jsonDict,
                false
            );

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            try {
                Util::executeApiRequest(
                    'POST',
                    'https://api.sandbox.dana.id/rest/redirection/v1.0/debit/payment-host-to-host',
                    $headers,
                    $jsonDict
                );

                $this->fail('Expected ApiException for missing X-TIMESTAMP but the API call succeeded');
            } catch (ApiException $e) {
                Assertion::assertApiException(
                    self::$jsonPathFileWidget,
                    self::$titleCase,
                    $caseName,
                    $e,
                    [],
                    $jsonDict
                );
            }
        });
    }

    /**
     * @testdox Should give fail response for invalid signature scenario
     * @description Validates security by testing invalid signature rejection
     * @test Verifies that requests with invalid signatures are properly rejected
     * @expectedResult API returns signature validation error with security-related message
     * @category SecurityValidation
     * @priority Critical
     * @author Integration Test Team
     * @since 1.0.0
     */
    public function testPaymentInvalidSignature(): void
    {
        Util::withDelay(function () {
            $caseName = 'PaymentFailInvalidSignature';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );

            $jsonDict['merchantId'] = self::$merchantId;
            $jsonDict['partnerReferenceNo'] = PaymentUtil::generatePartnerReferenceNo();
            $jsonDict['validUpTo'] = Util::generateFormattedDate(600, 7);

            // Create headers without timestamp to test validation
            $headers = Util::getHeadersWithSignature(
                'POST',
                '/rest/redirection/v1.0/debit/payment-host-to-host.htm',
                $jsonDict,
                true,
                false,
                true
            );

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            try {
                Util::executeApiRequest(
                    'POST',
                    'https://api.sandbox.dana.id/rest/redirection/v1.0/debit/payment-host-to-host',
                    $headers,
                    $jsonDict
                );

                $this->fail('Expected ApiException for missing X-TIMESTAMP but the API call succeeded');
            } catch (Exception $e) {
                Assertion::assertApiException(
                    self::$jsonPathFileWidget,
                    self::$titleCase,
                    $caseName,
                    $e,
                    [],
                    $jsonDict
                );
            }
        });
    }

    /**
     * @testdox Should give fail response for timeout scenario
     * @description Validates timeout handling when API requests exceed time limits
     * @test Verifies that timeout scenarios return appropriate HTTP 504 Gateway Timeout
     * @expectedResult API returns 504 Gateway Timeout error code for timeout scenarios
     * @category TimeoutHandling
     * @priority Medium
     * @author Integration Test Team
     * @since 1.0.0
     */
    public function testPaymentTimeout(): void
    {
        $this->markTestSkipped('Skipping testPaymentTimeout as requested.');
        Util::withDelay(function () {
            $caseName = 'PaymentFailTimeout';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );

            $jsonDict['merchantId'] = self::$merchantId;
            $jsonDict['partnerReferenceNo'] = PaymentUtil::generatePartnerReferenceNo();
            $jsonDict['validUpTo'] = Util::generateFormattedDate(600, 7);

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            try {
                self::$apiInstanceWidget->widgetPayment($requestObj);
            } catch (ApiException $e) {
                $this->assertEquals(504, $e->getCode(), "Gateway Time-out");
            }
        });
    }

    /**
     * @testdox Should verify idempotency behavior for payment requests
     * @description Validates idempotency behavior by testing duplicate request handling
     * @test Verifies that duplicate requests with same reference return consistent responses
     * @expectedResult Duplicate requests either return identical responses or proper duplicate errors
     * @category IdempotencyValidation
     * @priority High
     * @author Integration Test Team
     * @since 1.0.0
     * @steps
     * 1. Make initial payment request with fixed partner reference
     * 2. Make immediate duplicate request with same reference
     * 3. Make delayed duplicate request to test persistence
     * @scenarios
     * - True Idempotency: Returns identical response objects
     * - Duplicate Detection: Returns proper duplicate error messages
     */
    public function testPaymentIdempotent(): void
    {
        $this->markTestSkipped('Skipping testPaymentIdempotent as requested.');
        Util::withDelay(function () {
            $caseName = 'PaymentIdempotent';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );

            $jsonDict['merchantId'] = self::$merchantId;
            $fixedPartnerRef = 'IDEMPOTENT_TEST_' . time() . '_' . substr(md5(rand()), 0, 8);
            $jsonDict['partnerReferenceNo'] = $fixedPartnerRef;
            $jsonDict['validUpTo'] = Util::generateFormattedDate(600, 7);

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            $isIdempotent = false;
            $firstResponse = null;

            try {
                // STEP 1: First request - should succeed
                $firstResponse = self::$apiInstanceWidget->widgetPayment($requestObj);

                sleep(1);

                // STEP 2: Second request with SAME partnerReferenceNo - test idempotency
                try {
                    $secondResponse = self::$apiInstanceWidget->widgetPayment($requestObj);

                    if ($firstResponse->getReferenceNo() === $secondResponse->getReferenceNo()) {
                        $isIdempotent = true;

                        $this->assertEquals(
                            $firstResponse->getReferenceNo(),
                            $secondResponse->getReferenceNo(),
                            'Reference numbers must be identical for truly idempotent requests'
                        );

                        if (method_exists($firstResponse, 'getResponseCode') && method_exists($secondResponse, 'getResponseCode')) {
                            $this->assertEquals(
                                $firstResponse->getResponseCode(),
                                $secondResponse->getResponseCode(),
                                'Response codes must be identical for truly idempotent requests'
                            );
                        }
                    } else {
                        $this->fail('Both requests succeeded but returned different reference numbers - NOT truly idempotent');
                    }
                } catch (ApiException $duplicateError) {
                    $errorResponse = json_decode($duplicateError->getResponseBody(), true);

                    $isDuplicateError =
                        (isset($errorResponse['responseCode']) &&
                            (strpos($errorResponse['responseCode'], 'DUPLICATE') !== false ||
                                strpos($errorResponse['responseCode'], 'ALREADY_EXIST') !== false ||
                                strpos($errorResponse['responseCode'], 'IDEMPOTENT') !== false)) ||
                        (isset($errorResponse['responseMessage']) &&
                            (strpos(strtolower($errorResponse['responseMessage']), 'duplicate') !== false ||
                                strpos(strtolower($errorResponse['responseMessage']), 'already exist') !== false ||
                                strpos(strtolower($errorResponse['responseMessage']), 'idempotent') !== false));

                    if ($isDuplicateError) {
                        Assertion::assertApiException(
                            self::$jsonPathFileWidget,
                            self::$titleCase,
                            $caseName,
                            $duplicateError,
                            ['partnerReferenceNo' => $fixedPartnerRef],
                            $jsonDict
                        );
                    } else {
                        $this->fail('Expected duplicate error, but got unexpected error: ' . $duplicateError->getMessage());
                    }
                }

                // STEP 3: Test with longer delay to check idempotency persistence
                sleep(1);

                $thirdResponse = self::$apiInstanceWidget->widgetPayment($requestObj);

                if ($isIdempotent) {
                    Assertion::assertResponse(
                        self::$jsonPathFileWidget,
                        self::$titleCase,
                        $caseName,
                        $thirdResponse->__toString(),
                        []
                    );
                }
            } catch (ApiException $firstError) {
                $this->markTestSkipped('First request failed - cannot test idempotency: ' . $firstError->getMessage());
            }
        });
    }

    /**
     * @testdox Should give fail response for internal server error scenario
     * @description Validates error handling when server encounters internal processing errors
     * @test Verifies that server-side errors return appropriate HTTP 500 Internal Server Error
     * @expectedResult API returns 500 Internal Server Error code for server-side processing failures
     * @category ServerErrorHandling
     * @priority Medium
     * @author Integration Test Team
     * @since 1.0.0
     */
    public function testPaymentFailInternalServerError(): void
    {
        Util::withDelay(function () {
            $caseName = 'PaymentFailInternalServerError';
            $jsonDict = Util::getRequest(
                self::$jsonPathFileWidget,
                self::$titleCase,
                $caseName
            );

            $jsonDict['merchantId'] = self::$merchantId;
            $jsonDict['partnerReferenceNo'] = PaymentUtil::generatePartnerReferenceNo();
            $jsonDict['validUpTo'] = Util::generateFormattedDate(600, 7);

            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\WidgetPaymentRequest'
            );

            try {
                self::$apiInstanceWidget->widgetPayment($requestObj);
                $this->fail('Expected ApiException for internal server error but the API call succeeded');
            } catch (ApiException $e) {
                // Verify it's a 500 Internal Server Error
                $this->assertEquals(500, $e->getCode(), "Expected HTTP 500 Internal Server Error, got {$e->getCode()}");
                
                // Use assertApiException to validate the error response
                Assertion::assertApiException(
                    self::$jsonPathFileWidget,
                    self::$titleCase,
                    $caseName,
                    $e,
                    [],
                    $jsonDict
                );
            } catch (Exception $e) {
                $this->fail('Unexpected exception: ' . $e->getMessage());
            }
        });
    }
}

File: test/python/widget/payment_test.py
import os
import pytest
import threading
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from dana.utils.snap_configuration import SnapConfiguration, AuthSettings, Env
from dana.widget.v1.enum import *
from dana.widget.v1.models import *
from dana.widget.v1 import *
from dana.widget.v1.api import *
from dana.api_client import ApiClient
from dana.exceptions import *
from helper.util import get_request, with_delay, generate_partner_reference_no, retry_on_inconsistent_request, retry_on_inconsistent_request
from helper.assertion import assert_response, assert_fail_response
from helper.api_helpers import get_headers_with_signature, execute_and_assert_api_error

title_case = "Payment"
json_path_file = "resource/request/components/Widget.json"
merchant_id = os.environ.get("MERCHANT_ID", "default_merchant_id")

configuration = SnapConfiguration(
    api_key=AuthSettings(
        PRIVATE_KEY=os.environ.get("PRIVATE_KEY"),
        ORIGIN=os.environ.get("ORIGIN"),
        X_PARTNER_ID=os.environ.get("X_PARTNER_ID"),
        CLIENT_SECRET=os.environ.get("CLIENT_SECRET"),
        ENV=Env.SANDBOX,
        DANA_ENV=Env.SANDBOX,
        X_DEBUG="true"
    )
)

with ApiClient(configuration) as api_client:
    api_instance = WidgetApi(api_client)


@with_delay()
@retry_on_inconsistent_request(max_retries=3,delay_seconds=2)
def test_payment_success():
    case_name = "PaymentSuccess"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    partner_reference_no = generate_partner_reference_no()
    json_dict["partnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id
    
    # Convert the request data to a CreateOrderRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)
    api_response = api_instance.widget_payment(create_payment_request_obj)

    # Assert the API response
    assert_response(json_path_file, title_case, case_name, WidgetPaymentResponse.to_json(api_response), {"partnerReferenceNo": partner_reference_no})

@with_delay()
def test_payment_fail_invalid_format():
    case_name = "PaymentFailInvalidFormat"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    partner_reference_no = generate_partner_reference_no()
    json_dict["partnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id
    
    # Convert the request data to a CreateOrderRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)
    try:
        api_response = api_instance.widget_payment(create_payment_request_obj)
        pytest.fail("Expected BadRequestException but the API call succeeded")
    except BadRequestException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": partner_reference_no})
    except:
        pytest.fail("Expected BadRequestException but the API call give another exception")

@with_delay()
def test_payment_fail_missing_or_invalid_mandatory_field():
    case_name = "PaymentFailMissingOrInvalidMandatoryField"
    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
    
    # Set a unique partner reference number
    partner_reference_no = generate_partner_reference_no()
    json_dict["partnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a CreateOrderRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)
    
    # Prepare headers without timestamp to trigger mandatory field error
    # This now handles the signature generation internally
    headers = get_headers_with_signature(
        method="POST",
        resource_path="/rest/redirection/v1.0/debit/payment-host-to-host",
        request_obj=json_dict,
        with_timestamp=False
    )
    
    # Execute the API request and assert the error
    execute_and_assert_api_error(
        api_client,
        "POST",
        "http://api.sandbox.dana.id/rest/redirection/v1.0/debit/payment-host-to-host",
        create_payment_request_obj,
        headers,
        400,  # Expected status code
        json_path_file,
        title_case,
        case_name,
        {"partnerReferenceNo": partner_reference_no}
    )

@with_delay()
def test_payment_fail_invalid_signature():
    case_name = "PaymentFailInvalidSignature"
    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
    
    # Set a unique partner reference number
    partner_reference_no = generate_partner_reference_no()
    json_dict["partnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id
    
    # Convert the request data to a CreateOrderRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)

    # Prepare headers with invalid signature to trigger authorization error
    # Since we're only using the invalid signature flag, we don't need to pass any of the other parameters
    headers = get_headers_with_signature(invalid_signature=True)
    
    # Execute the API request and assert the error
    execute_and_assert_api_error(
        api_client,
        "POST",
        "http://api.sandbox.dana.id/rest/redirection/v1.0/debit/payment-host-to-host",
        create_payment_request_obj,
        headers,
        401,  # Expected status code
        json_path_file,
        title_case,
        case_name,
        {"partnerReferenceNo": partner_reference_no}
    )

@with_delay()
def test_payment_fail_merchant_not_exist_or_status_abnormal():
    case_name = "PaymentFailMerchantNotExistOrStatusAbnormal"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    partner_reference_no = generate_partner_reference_no()
    json_dict["partnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id
    
    # Convert the request data to a CreateOrderRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)
    try:
        api_response = api_instance.widget_payment(create_payment_request_obj)
        pytest.fail("Expected NotFoundException but the API call succeeded")
    except NotFoundException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": partner_reference_no})
    except:
        pytest.fail("Expected NotFoundException but the API call give another exception")

@with_delay()
def test_payment_fail_inconsistent_request():
    case_name = "PaymentFailInconsistentRequest"

    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)

    # Set the partner reference number
    partner_reference_no = generate_partner_reference_no()
    json_dict["originalPartnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id

    # Convert the request data to a RefundOrderRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)

    # Make the API call and assert the response
    try:
        # First hit API
        api_instance.widget_payment(create_payment_request_obj)
        create_payment_request_obj.amount = Money(currency="IDR", value="10000.00")
        # Second hit API with same data (duplicate)
        api_instance.widget_payment(create_payment_request_obj)
    except NotFoundException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": json_dict["originalPartnerReferenceNo"]})

@with_delay()
def test_payment_fail_internal_server_error():
    case_name = "PaymentFailInternalServerError"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    partner_reference_no = generate_partner_reference_no()
    json_dict["partnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id
    
    # Convert the request data to a CreateOrderRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)
    try:
        api_response = api_instance.widget_payment(create_payment_request_obj)
        pytest.fail("Expected ServiceException but the API call succeeded")
    except ServiceException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body, {"partnerReferenceNo": partner_reference_no})
    except:
        pytest.fail("Expected ServiceException but the API call give another exception")

@with_delay()
def test_payment_fail_timeout():
    case_name = "PaymentFailTimeout"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    partner_reference_no = generate_partner_reference_no()
    json_dict["partnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id
    
    # Convert the request data to a CreateOrderRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)
    try:
        api_instance.widget_payment(create_payment_request_obj)
        pytest.fail("Expected ServiceException but the API call succeeded")
    except ServiceException as e:
        assert e.status == 504, "Expected status code 504 for timeout"
    except:
        pytest.fail("Expected ServiceException but the API call give another exception")  
        assert_fail_response(json_path_file, title_case, case_name, e.body, None)

@with_delay()
def test_payment_idempotent():
    """Test payment idempotent - same request should return same result"""
    case_name = "PaymentIdempotent"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    partner_reference_no = generate_partner_reference_no()
    json_dict["partnerReferenceNo"] = partner_reference_no
    json_dict["merchantId"] = merchant_id
    
    # Convert the request data to a WidgetPaymentRequest object
    create_payment_request_obj = WidgetPaymentRequest.from_dict(json_dict)
    
    # First hit API - should succeed
    first_response = api_instance.widget_payment(create_payment_request_obj)
    # Second hit API with same data - should return same result (idempotent)
    second_response = api_instance.widget_payment(create_payment_request_obj)
    
    # Verify idempotent behavior - both responses should be identical
    assert first_response.partner_reference_no == second_response.partner_reference_no, "Partner reference numbers should match for idempotent requests"
    
    # Assert the response
    assert_response(json_path_file, title_case, case_name, WidgetPaymentResponse.to_json(second_response), {"partnerReferenceNo": partner_reference_no})